File: 211. Design Add and Search Words Data Structure/main.py
class WordDictionary:
    def __init__(self):
        self.children = dict()
        self.isEndOfWord = False

    def addWord(self, word: str) -> None:
        node = self

        for c in word:
            if c not in node.children:
                node.children[c] = WordDictionary()
            node = node.children[c]

        node.isEndOfWord = True

    def search(self, word: str) -> bool:
        node = self
        wordLen = len(word)

        for i in range(wordLen):
            if word[i] == ".":
                for wd in node.children.values():
                    if wd.search(word[i + 1 :]):
                        return True

                return False

            if word[i] not in node.children:
                return False

            node = node.children[word[i]]

        return node.isEndOfWord


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: 211. Design Add and Search Words Data Structure/main.cpp
#include <string>
#include <unordered_map>

using namespace std;

class WordDictionary {
    static constexpr size_t ALPHABET_SIZE = 26;

    unordered_map<char, WordDictionary *> children;
    bool isEndOfWord;

public:
    WordDictionary() {
        isEndOfWord = false;
    }

    void addWord(string word) {
        WordDictionary *node = this;

        for (const auto &c : word) {
            if (!node->children.contains(c)) {
                node->children[c] = new WordDictionary();
            }
            node = node->children[c];
        }

        node->isEndOfWord = true;
    }

    bool search(string word) {
        WordDictionary *node = this;
        size_t wordLen = word.size();

        for (size_t i = 0; i < wordLen; i++) {
            if (word[i] == '.') {
                for (const auto &[c, wd] : node->children) {
                    if (wd->search(word.substr(i + 1, wordLen))) {
                        return true;
                    }
                }

                return false;
            }

            if (!node->children.contains(word[i])) {
                return false;
            }
            node = node->children[word[i]];
        }

        return node->isEndOfWord;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */